File: btypes/src/main/java/de/hhu/stups/btypes/BInteger.java
package de.hhu.stups.btypes;

/**
 * Created by fabian on 15.10.18.
 */
public class BInteger extends BNumber {


    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (obj instanceof BNumber) {
            //TODO: other numbers
            return this.compareTo((BNumber) obj) == 0;
        }
        // assert getClass() != obj.getClass()
        return false;
    }

    private static final long serialVersionUID = -6484548796859331267L;
    private int value;

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + value;
        return result;
    }

    public BInteger(int value) {
        this.value = value;
    }

    public int compareTo(BNumber o) {
        BInteger other = (BInteger) o;
        return this.value - other.value;
    }

    public BBoolean lessEqual(BNumber o) {
        return new BBoolean(compareTo(o) <= 0);
    }


    public BBoolean greaterEqual(BNumber o) {
        return new BBoolean(compareTo(o) >= 0);
    }

    @Override
    public java.math.BigInteger asBigInteger() {
        return new java.math.BigInteger(String.valueOf(value));
    }

    public BBoolean less(BNumber o) {
        return new BBoolean(compareTo(o) < 0);
    }

    public BBoolean greater(BNumber o) {
        return new BBoolean(compareTo(o) > 0);
    }

    public BBoolean equal(BNumber o) {
        return new BBoolean(compareTo(o) == 0);
    }

    public BBoolean unequal(BNumber o) {
        return new BBoolean(compareTo(o) != 0);
    }

    @Override
    public int intValue() {
        return this.value;
    }

    @Override
    public long longValue() {
        return (long) this.value;
    }

    @Override
    public float floatValue() {
        return (float) this.value;
    }

    @Override
    public double doubleValue() {
        return (double) this.value;
    }

    @Override
    public BNumber plus(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value + other.value);
    }

    public java.lang.String toString() {
        return String.valueOf(value);
    }

    @Override
    public BNumber minus(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value - other.value);
    }

    @Override
    public BNumber multiply(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value * other.value);
    }

    @Override
    public BNumber power(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value ^ other.value);
    }

    @Override
    public BNumber divide(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value / other.value);
    }

    @Override
    public BNumber modulo(BNumber o) {
        BInteger other = (BInteger) o;
        return new BInteger(this.value % other.value);
    }

    @Override
    public BNumber or(BNumber o) {
        return null;
    }

    @Override
    public BNumber and(BNumber o) {
        return null;
    }

    @Override
    public BNumber xor(BNumber o) {
        return null;
    }

    @Override
    public BNumber next() {
        return new BInteger(this.value + 1);
    }

    @Override
    public BNumber previous() {
        return new BInteger(this.value - 1);
    }

    @Override
    public BNumber leftShift(BNumber o) {
        return new BInteger(this.value << 1);
    }

    @Override
    public BNumber rightShift(BNumber o) {
        return new BInteger(this.value >> 1);
    }

    @Override
    public boolean isCase(BNumber o) {
        return this.equals(o);
    }

    @Override
    public BNumber negative() {
        return new BInteger(-this.value);
    }

    @Override
    public BNumber positive() {
        return this;
    }

    public int getValue() {
        return value;
    }

}

File: codegenerator/src/test/resources/de/hhu/stups/codegenerator/BInteger.cpp
#include <iostream>
#include <string>
#include "BBoolean.cpp"

#ifndef BINTEGER_H
#define BINTEGER_H

using namespace std;

class BInteger : public BObject {

    /*
	@Override
	public boolean equals(Object obj) {
		if (== obj)
			return true;
		if (obj == null)
			return false;
		if (obj instanceof java.lang.Number) {
			return compareTo((java.lang.Number) obj) == 0;
		}
		// assert getClass() != obj.getClass()
		return false;
	}*/

	private:
	    int value;

	/*@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((value == null) ? 0 : value.hashCode());
		return result;
	}*/

    public:
        /*BInteger(string val) {
            value = new java.math.BigInteger(val);
        }*/

        BInteger(int val) {
            value = val;
        }

        BInteger(BInteger* val) {
            this->value = val->value;
        }

        int intValue() const {
            return value;
        }

        /*int compareTo(BInteger o) {
            return value.compareTo(o.asBigInteger());
        }*/

        int compareTo(BInteger* o) const {
            return intValue() - o->intValue();
        }

        BBoolean* lessEqual(BInteger* o) {
            return new BBoolean(compareTo(o) <= 0);
        }


        BBoolean* greaterEqual(BInteger* o) {
            return new BBoolean(compareTo(o) >= 0);
        }

        BBoolean* less(BInteger* o) {
            return new BBoolean(compareTo(o) < 0);
        }

        BBoolean* greater(BInteger* o) {
            return new BBoolean(compareTo(o) > 0);
        }

        BBoolean* equal(BInteger* o) {
            return new BBoolean(compareTo(o) == 0);
        }

        BBoolean* unequal(BInteger* o) {
            return new BBoolean(compareTo(o) != 0);
        }

        /*int compareTo(java.lang.Number o) {
            java.math.BigInteger oi;
            if (o == null) {
                throw new NullPointerException();
            }
            if (getClass() != o.getClass()) {
                oi = new java.math.BigInteger(java.lang.String.valueOf(o.longValue()));
            } else {
                BInteger oo = (BInteger) o;
                oi = oo.value;
            }
            return value.compareTo(oi);
        }*/

        /*long longValue() {
            return value.longValue();
        }

        float floatValue() {
            return value.floatValue();
        }

        double doubleValue() {
            return value.doubleValue();
        }*/

        BInteger* plus(BInteger* o) {
            return new BInteger(value + o->intValue());
        }

        /*java.math.BigInteger asBigInteger() {
            return value;
        }

        java.lang.String toString() {
            return value.toString();
        }*/

        BInteger* minus(BInteger* o) {
            return new BInteger(value - o->intValue());
        }

        BInteger* multiply(BInteger* o) {
            return new BInteger(value * o->intValue());
        }

        BInteger* power(BInteger* o) {
            return new BInteger(value ^ o->intValue());
        }

        BInteger* divide(BInteger* o) {
            return new BInteger(value / o->intValue());
        }

        BInteger* modulo(BInteger* o) {
            return new BInteger(value % o->intValue());
        }

        /*BInteger or(BInteger o) {
            return new BInteger(value.or(o.value));
        }

        BInteger and(BInteger o) {
            return new BInteger(value.and(o.value));
        }

        BInteger xor(BInteger o) {
            return new BInteger(value.xor(o.value));
        }*/

        BInteger* next() {
            return new BInteger(value + 1);
        }

        BInteger* previous() {
            return new BInteger(value - 1);
        }

        /*BInteger leftShift(BInteger o) {
            return new BInteger(value.shiftLeft(o.intValue()));
        }

        BInteger rightShift(BInteger o) {
            return new BInteger(value.shiftRight(o.intValue()));
        }

        boolean isCase(BInteger o) {
            return equals(o);
        }*/

        BInteger* negative() {
            return new BInteger(-value);
        }

        BInteger* positive() {
            return this;
        }

        BInteger operator=(BInteger other) {
            this->value = other.value;
            return *this;
        }

        friend bool operator !=(const BInteger& o1, const BInteger& o2) {
            return o1.intValue() - o2.intValue() != 0;
        }

        friend bool operator ==(const BInteger& o1, const BInteger& o2) {
            std::cout << typeid(o1).name() << std::endl;
            return o1.intValue() - o2.intValue() == 0;
        }

        friend bool operator <(const BInteger& p1, const BInteger& p2) {
            return p1.intValue() < p2.intValue();
        }

        int hashCode() const override {
            int prime = 31;
            int result = 1;
            result = prime * result + value;
            return result;
        }

};
#endif